sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala

```diff
@@ -193,88 +144,208 @@ object CollationTypeCoercion {
     case other => other
   }
 
+  /**
+   * If childType is collated and target is UTF8_BINARY, the collation of the output
+   * should be that of the childType.
+   */
+  private def shouldRemoveCast(cast: Cast): Boolean = {
+    val isUserDefined = cast.getTagValue(Cast.USER_SPECIFIED_CAST).isDefined
+    val isChildTypeCollatedString = cast.child.dataType match {
```

Comment:
do we need this condition? If the child is utf8 collation, we can remove the cast as well as the target type is also utf8 collation.

Reply:
Actually I remember now why I did it, doing it for all strings will create a huge diff in analyzed plan golden files for non-collated stuff, and that cast will get removed in the optimizer anyways, so I would rather keep it.